### PR TITLE
Use Project API for Ballerina to OpenAPI code generation

### PR DESCRIPTION
**Pull request by lnash94**
https://github.com/wso2-enterprise/internal-support-ballerina/issues/275 

While doing implementation we encounter that when using `Yaml.pretty()` to generate YAML output in the Swagger YAML utility, the output can differ between Linux and Windows. 
output for Linux
```openapi
openapi: 3.0.1
info:
  title: PayloadV
  version: 0.1.0
servers:
  - url: "{server}:{port}/payloadV"
    variables:
      server:
        enum:
          - http://petstore.openapi.io
          - http://localhost
        default: http://localhost
      port:
        enum:
          - "443"
          - "80"
        default: "80"
paths:
...
```
output for Windows
```openapi
openapi: 3.0.1
info:
  title: PayloadV
  version: 0.1.0
servers:
  - url: "{server}:{port}/payloadV"
    variables:
      server:
        default: http://localhost
        enum:
          - http://petstore.openapi.io
          - http://localhost
      port:
        default: "80"
        enum:
          - "443"
          - "80"   
paths:
...
```
The reason can be differences in how these operating systems handle line breaks.
Therefore we decided to add both Windows and Linux tests since the both content is the same and only the order may change. This won't be a matter for accessing data from generated OpenAPI specifications.

This project API change affected to 
- The Baalerina to OpenAPI spec generation in CLI
- Http openAPI spec generation extensions
- Ballerina build command `--export-openapi` flag
- LS extension for openAPI spec generations

### CODE REVIEW

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/model/OASGenerationMetaInfo.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ */
+public class OASGenerationMetaInfo {
+
+    private final String openApiFileName;
+    private final Path ballerinaFilePath;
+    private final SemanticModel semanticModel;
+    private final ServiceDeclarationNode serviceDeclarationNode;
+    private final LinkedHashSet<ListenerDeclarationNode> endpoints;
+    //TODO: enable when complete constraint support
+//    private final Set<TypeDefinitionNode> typeDefinitionNodes;
```
Comment (NipunaRanasinghe) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/model/OASGenerationMetaInfo.java:41]: lets format this properly

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/model/OASGenerationMetaInfo.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    public ServiceDeclarationNode getServiceDeclarationNode() {
+        return serviceDeclarationNode;
+    }
+
+    public LinkedHashSet<ListenerDeclarationNode> getEndpoints() {
+        return endpoints;
+    }
+
```
Comment (NipunaRanasinghe) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/model/OASGenerationMetaInfo.java:71]: Additional blank line

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/model/OASGenerationMetaInfo.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        private String openApiFileName;
+        private Path ballerinaFilePath;
+        private SemanticModel semanticModel;
+        private ServiceDeclarationNode serviceDeclarationNode;
+        private LinkedHashSet<ListenerDeclarationNode> endpoints;
+//        private Set<TypeDefinitionNode> typeDefinitionNodes;
+
+        public OASGenerationMetaInfoBuilder setBallerinaFilePath(Path ballerinaFilePath) {
+
```
Comment (NipunaRanasinghe) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/model/OASGenerationMetaInfo.java:86]: Additional blank line

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/model/OASGenerationMetaInfo.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            this.ballerinaFilePath = ballerinaFilePath;
+            return this;
+        }
+
+        public OASGenerationMetaInfoBuilder setSemanticModel(SemanticModel semanticModel) {
+
```
Comment (NipunaRanasinghe) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/model/OASGenerationMetaInfo.java:92]: Additional blank line

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/ServiceToOpenAPIConverterUtils.java [hunk 5]
```diff
@@ -191,44 +205,27 @@ private static void extractListenersAndServiceNodes(String serviceName, List<Str
...
+                oasGenerationMetaInfo.getEndpoints(),
+                oasGenerationMetaInfo.getSemanticModel(), oasGenerationMetaInfo.getOpenApiFileName(),
+                oasGenerationMetaInfo.getBallerinaFilePath());
+    }
+
+    private static OASResult ensembleOASResult(ServiceDeclarationNode serviceDefinition,
+                                               LinkedHashSet<ListenerDeclarationNode> endpoints,
+                                               SemanticModel semanticModel,
+                                               String openApiFileName, Path ballerinaFilePath) {
+
```
Comment (NipunaRanasinghe) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/ServiceToOpenAPIConverterUtils.java:228]: Do we need 2 methods for this? Can't we just extract the `oasGenerationMetaInfo` within the same function and use it?

Comment (lnash94) [reply openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/ServiceToOpenAPIConverterUtils.java]: Yes, we can implement it within one function.

**Review by ayeshLK — APPROVED** [review]

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/BallerinaNodeVisitor.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Visitor to get the TypeDefinitionNode and ListenerDeclarationNodes.
+ *
+ * @since 1.6.0
+ */
+public class BallerinaNodeVisitor extends NodeVisitor {
```
[the remark this answers is not in the document]
Comment (lnash94) [reply openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/BallerinaNodeVisitor.java:33]: will use `ModuleMemberVisitor` for the visitor

**Review by aneeshafedo — APPROVED** [review]

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

**Pull request merged**